Add RuntimeError in _distn_infrastructure.py in fit() method

rlucas7 opened the pull request:
This PR is _one_ proposal to fix #10300 from the discussion w/ @chrisb83  @WarrenWeckesser  and  @kmundnic (who opened the issue). From the discussion on the issue thread it is not clear to me if we want to implement a `nan_policy` for fit or simply throw an error. Given the inconsistency in the current behavior any fix will change the behavior of the `fit()` method. I'd rather have the fix w/minimal disruption to the user base, not sure if this approach or `nan_policy` has less disruption. This implementation was quicker to put together than a full `nan_policy`. However, if we want a full `nan_policy` we should close this PR w/o merging and I'll work towards implementing a `nan_policy`. The 3 commits here are minimal, and replicate the issue in #10300 and resolve the issue by throwing a `RuntimeError` with a descriptive message of the issue.

chrisb83 commented on scipy/stats/_distn_infrastructure.py:
```diff
@@ -2256,6 +2258,9 @@ def fit(self, data, *args, **kwds):
             raise TypeError("Too many input arguments.")
 
+        if np.logical_not(np.isfinite(data)).any():
+            raise RuntimeError("The data contains np.nan values.")
```
"contains non-finite values" is better

rlucas7 replied:
changes the message in error on recent commits (in all occurrences)

chrisb83 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1412,6 +1412,11 @@ def get_moms(lam, sig, mu):
         assert_almost_equal(sts, get_moms(lam, sig, mu))
 
+    def test_npnan_raises_error(self):
+        with pytest.raises(RuntimeError) as excinfo:
```
you can use `assert_raises`, that is aligned to other tests

Maybe add another test for `np.inf`?

rlucas7 replied:
added a test for `np.inf`

chrisb83 commented:
> it is not clear to me if we want to implement a nan_policy for fit or simply throw an error.

My impression was that throwing an error was the prefered option. (`nan_policy` has quite a few issues)

WarrenWeckesser commented:
It will probably be a week or so before I can give a real review of this.  For now I'll just note that several of the distributions (e.g. `norm`, `beta` and a few others) override the `fit` method, so they will have to be updated individually.

rlucas7 commented:
>My impression was that throwing an error was the prefered option. (nan_policy has quite a few issues)

@chrisb83 after reading again it is clear to me the consensus on throwing an error. Not sure why I hadn't caught that before.

>It will probably be a week or so before I can give a real review of this. For now I'll just note that several of the distributions (e.g. norm, beta and a few others) override the fit method, so they will have to be updated individually.

I know you're busy @WarrenWeckesser and your comments are always helpful. I added some lines into the overridden `.fit()` methods and added full distributional unit test coverage.

rlucas7 commented:
@chrisb83 could we also add a 1.4.0 release tag on this so we at least revisit before the next release? (hopefully gets in sooner).

tylerjereddy commented:
> could we also add a 1.4.0 release tag on this so we at least revisit before the next release? 

Yes, done

rlucas7 commented:
> > could we also add a 1.4.0 release tag on this so we at least revisit before the next release?
> 
> Yes, done

Thanks @tylerjereddy !

pv commented on scipy/stats/_continuous_distns.py:
```diff
@@ -264,6 +264,9 @@ def fit(self, data, **kwds):
 
         data = np.asarray(data)
 
+        if np.logical_not(np.isfinite(data)).any():
```
Drive-by-comment: `not np.isfinite(data).all()` is likely faster since `logical_not` generates an intermediate array.

rlucas7 replied:
> Drive-by-comment: `not np.isfinite(data).all()` is likely faster since `logical_not` generates an intermediate array.

I bet this matters more on larger arrays. I was surprised to learn even on the test case there is a speedup: 
```
import scipy.stats as stats
import numpy as np
from datetime import datetime
startTime = datetime.now()
for i in  range(0, 10000):
    x = np.array([1.6483, 2.7169, 2.4667, 1.1791, 3.5433, np.nan])          
    np.testing.assert_raises(RuntimeError, stats.norm.fit, x)

print(datetime.now() - startTime)
```
gives runtime of: 
# 0:00:00.156653 (old) 
# 0:00:00.138408 (proposed)

for a larger range 10000 -> 1000000 the improvements are: 
0:00:12.887804 (old) 
0:00:12.201071 (new)   
 
I'll change all the implementations to that proposed by @pv  Thanks!

rlucas7 commented:
I just rebased the branch off master, commit `2b41710`, the branch was pretty stale (~662 commits behind master). There was only 1 conflict in the `TestFitMethod` looks like some things were either removed or factored out to a different file. After the conflict resolutions the `stats` tests all pass on my local.

rlucas7 commented:
> I just rebased the branch off master, commit `2b41710`, the branch was pretty stale (~662 >commits behind master). There was only 1 conflict in the `TestFitMethod` looks like some things >were either removed or factored out to a different file. After the conflict resolutions the `stats` >tests all pass on my local. 
>Would be great to get this into the 1.4.0 release.

I also added a backport candidate label. I'm still unclear on what items we usually backport. This one seems like a good candidate b/c it makes inconsistent behavior in the `fit()` methods consistent. If the committer thinks it isn't a backport candidate that's fine too but figured I'd label so we consider it for the backport.

tylerjereddy commented:
>  I'm still unclear on what items we usually backport

Usually just bug fixes--is this bad enough to be considered a bug?

rlucas7 commented:
> > I'm still unclear on what items we usually backport
> 
> Usually just bug fixes--is this bad enough to be considered a bug?

I'm not sure, the behavior on some fit methods if `nan` or `inf` is part of fitting data is inconsistent across  distributions. 
Sounds like this isn't a bug, it isn't too serious, whatever behavior a user is seeing they've likely been seeing for awhile.
I'll remove the backport label, thanks for clarifying @tylerjereddy

tylerjereddy commented:
@WarrenWeckesser Any concerns here? It looks like you were planning to review but didn't get around to it. Beyond that, it seems reviewer comments were addressed.

rlucas7 commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -1842,6 +1888,22 @@ class TestFitMethod(object):
+        """gh-10300"""
+        if dist in self.fitSkipNonFinite:
+            pytest.skip("%s fit known to fail or depracated" % dist)
+        with np.errstate(all='ignore'), suppress_warnings() as sup:
```
@WarrenWeckesser thanks! You are right, the context manager is not needed. This was a copy paste error IIRC. Doesn't add anything, removed. If the tests all pass I can do a rebase too, this branch is now fairly stale.

rlucas7 commented:
> +1 for the consistent handling of non-finite values. I'm OK with raising an exception. This leaves open the possibility of adding the `nan_policy` argument sometime in the future. One thing to be aware of: the change here raises an exception on `nan` or `inf`. `nan_policy` has been specifically about `nan`, not `inf`. I think expanding the `nan_policy` to include handling of `inf` in a consistent and backwards-compatible way is something we can do. I'm just noting that once we commit here to raising on `inf`, a future addition of a `nan_policy` argument must provide an option for raising errors on `inf` (along with `nan`). That is, we'll need an API for `nan_policy` that has an option for erroring on just `nan` or for erroring on any non-finite values, and the `nan_policy` for the `fit` method would need to use the option "raise on `nan` or `inf`" as its default.

I agree that we likely want to have something to handle `nan` and something to handle `inf`. 
I'm not sure that `nan_policy` is the best name for it though :). 
Maybe we could factor out an `inf_policy` and a `nan_policy`?
Seems a bit weird to label `nan_policy` if it is really `nan` ***and*** `inf` that are handled by the method. 
Maybe I'm splitting hairs here but it could be a `non_finite` policy which delegates to `nan_policy` and `inf_policy`.

rlucas7 commented:
Surprisingly no conflicts on rebase and perhaps my "largest" rebase yet:
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ git rev-list e90bafdfe1fa7266f3e3befc7f82368ce38bdf9f   2b417108907ff3254fe1 | wc -l 
   22226
```

rlucas7 commented:
@tylerjereddy @WarrenWeckesser I've removed the context manager as requested. Please let me know if you'd like anything else changed. 

We can take the `nan_policy` naming discussion to the dev-list.

WarrenWeckesser commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2020,6 +2066,20 @@ class TestFitMethod(object):
     def setup_method(self):
         np.random.seed(1234)
 
+    # skip these b/c depracated, or only loc and scale arguments
```
Spelling: deprecated

rlucas7 replied:
done

WarrenWeckesser commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -2020,6 +2066,20 @@ class TestFitMethod(object):
+    def test_fit_w_non_finite_data_values(self, dist, args):
+        """gh-10300"""
+        if dist in self.fitSkipNonFinite:
+            pytest.skip("%s fit known to fail or depracated" % dist)
```
Spelling: deprecated

rlucas7 replied:
done.

rlucas7 commented:
Thanks Warren. I added a brief note under backwards incompatible changes in the release notes wiki.
